### DISCUSSION TITLE
​Support additional Oauth2/OIDC/SAML providers

### DISCUSSION THREAD

**Discussion by dbrandonjohnson** [toeverything/AFFiNE#6473] [Ideas]
### Description

Add support for additional authentication providers that support Oauth2, OIDC or SAML such as [keycloak](https://www.keycloak.org/), [authentik](https://goauthentik.io/), Ipsilon or AzureAD.

Based on the current code base, it appears to only support Google and Github as identity providers.

### Use case

Adding additional authentication options provides authentication for Self-Hosted users and Affine Cloud business customers that use authentication providers other than Google Auth or Github.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

[a comment by an automated account is not shown]

**Comment by forehalo** [comment]
Other auth providers are truly on the table, but TBH, they are in low priority.

**Reply by VladimirLevadnij** [reply]
Hi. Tell me, please, is there a detailed discussion of this topic somewhere (oAuth2 / OIDC / SAML)? In particular, I am interested in how architecturally it is planned to be implemented, with the help of which libraries? Perhaps you plan to use https://www.passportjs.org / ?

**Reply by forehalo** [reply]
we haven't had any further discussion regarding this. There is one thing I can tell you is we won't use passportjs because the integration with nestjs, which we build the server with, is not considered good enough for us.

**Reply by VladimirLevadnij** [reply]
Thank you, please tell me if I understand correctly that it is planned to abandon nestjs in the future? Or are you looking to refactor an existing implementation using nestjs?

**Comment by otahirs** [comment]
I would not be looking for adding one or two additional providers. Instead I would focus on implementing generic support for oAuth2 / OIDC protocols. So any provider can be set-up without the need to revisit the issue.

**Reply by forehalo** [reply]
by providers I mean auth methods, not any specific oauth venders

**Comment by Felitendo** [comment]
I would really appreciate this because I need my friends and family to login with Authentik. Therefore adding custom OIDC/oAuth2 Providers as an option would be awesome :)

**Comment by llann** [comment]
This would be necessary to be able to use it for our company. Today, support for an external authentication system (SAML, OIDC, etc.) is essential, as managing users by hand in each of the tools is not possible for a company and also security wise. This would probably also be a feature to highlight in your paid packages ;)

**Comment by junkaiman** [comment]
Any updates on this?

I see there's OIDC OAuth provider config in the latest config system. But no clear documentation yet @forehalo

**Reply by junkaiman** [reply]
Just figured this out with the help of ChatGPT. I am using Authentik as an example.

1. Create an app with provider in Authentik
2. In the Provider settings, set redirect url to `https://affine.example.com/auth/callback`
3. In the "Advanced protocol settings" from the Provider setting, make sure to add 'offline_access' to the scope
4. Config in Affine admin portal:
```
{
  "clientId": "yyyy",
  "clientSecret": "xxxx",
  "issuer": "https://<your-authentik-instance-addr>/application/o/<your-app>/",
  "args": {
    "scope": "openid profile email offline_access"
  }
}
```
Save.

Seems like you have to save the runtime config every time the services get restarted

**Reply by Felitendo** [reply]
@junkaiman Thank you this also worked for me. Did you figure out how we can save the runtime config automatically?

**Reply by AllanChain** [reply]
> 2\. In the Provider settings, set redirect url to `https://affine.example.com/auth/callback`

The redirect URL should be <code>https://affine.example.com/<b>oauth</b>/callback</code>

**Reply by netliam** [reply]
Is there a launch url? I want to be able to automatically sign-in from my SSO dashboard.

**Reply by Daniel15** [reply]
You can leave the redirect URL blank in Authentik and it'll automatically set it correctly the first time you log in.

**Comment by drandarov-io** [comment]
https://docs.goauthentik.io/add-secure-apps/providers/oauth2/#email-scope-verification

`email_verified` is required and potentially default=false for Authentik 2025.10
